src/scores/processing/block_bootstrap.py

```diff
@@ -0,0 +1,410 @@
+
+    Args:
+        objects: The data to bootstrap, which can be multiple datasets. In the case where
+            multiple datasets are passed, each dataset can have its own set of dimension. However,
+            for successful bootstrapping, dimensions across all input objects must be nested.
+            For instance, for `block.keys=['d1', 'd2', 'd3'], an object with dimension 'd1' and
+            'd2' is valid, but an object with only dimension 'd2' is not valid. All datasets
+            are bootstrapped according to the same random samples along available dimensions.
+        blocks: A dictionary specifying the dimension(s) to bootstrap and the block sizes to
+            use along each dimension: `{dimension: block_size}`. The keys represent the dimensions
+            to be bootstrapped, and the values indicate the block sizes along each dimension.
+            The dimension provided here should exist in the data provided as `objects`.
+        n_iteration: The number of iterations to repeat the bootstrapping process. Determines
+            how many bootstrapped arrays will be generated and stacked along the iteration
+            dimension.
+        exclude_dims: An optional parameter indicating the dimensions to be excluded during
+            bootstrapping for each object provided in `objects`. This parameter expects a list
+            of lists, where each inner list corresponds to the dimensions to be excluded for
+            the respective object. By default, the assumption is that no dimensions are
+            excluded, and all objects are bootstrapped across all specified dimensions in `blocks`.
+        circular: A boolean flag indicating whether circular block bootstrapping should be
+            performed. Circular bootstrapping means that bootstrapping continues from the beginning
+            when the end of the data is reached. By default, this parameter is set to True.
+
+    Returns:
+        If a single Dataset/DataArray (XarrayLike) is provided, the functions returns a
+        bootstrapped XarrayLike object along the "iteration" dimension. If multiple XarrayLike
+        objects are provided, it returns a tuple of bootstrapped XarrayLike objects, each stacked
+        along the "iteration" dimension.
+
+    References:
+    Wilks, Daniel S. Statistical methods in the atmospheric sciences. Vol. 100.
+      Academic press, 2011.
+    """
+    # While the most efficient method involves expanding the iteration dimension withing the
+    # universal function, this approach might generate excessively large chunks (resulting
+    # from multiplying chunk size by iterations) leading to issues with large numbers of
+    # iterations. Hence, here function loops over blocks of iterations to generate the total
+    # number of iterations.
+
+    def _max_chunk_size_mb(ds):
+        """
+        Get the max chunk size in a dataset
+        """
+
+        def size_of_chunk(chunks, itemsize):
```

Comment:
[optional]

Consider replacing this whole thing with `numpy` e.g. `num = np.prod([np.max(x) for x in chunks])` (not tested...)